File: atmosphere.c
#include "ebm.h"

void calc_dif(const SimulationConfig *conf, const PlanetParams *para,
              ClimateState *state, double dif[]) {
  int reso = conf->reso;
  int n;
  double coefficient = 5.3e-3;
  double D = coefficient * state->P_air;
  double del_phi = M_PI / 180.0;
  double phi[reso];

  for (n = 1; n < reso - 1; n++) {
    phi[n] = (n - 90) * M_PI / 180.0;
    dif[n] =
        -D * tan(phi[n]) * (state->T[n + 1] - state->T[n - 1]) * 0.5 / del_phi +
        D * (state->T[n + 1] + state->T[n - 1] - 2.0 * state->T[n]) / del_phi /
            del_phi;
  }
  dif[0] = 0.0;
  dif[180] = 0.0;
}

void calc_radi(const SimulationConfig *conf, const PlanetParams *para,
               ClimateState *state, double radi[]) {
  int reso = conf->reso;
  int n;
  double x;
  double a[5], b[5];
  double A = 0.0, B = 0.0;
  double P_air = state->P_air;

  if (P_air > 7.45)
    P_air = 7.45;
  if (P_air < 3.4e-12)
    P_air = 3.4e-12;
  x = log10(P_air);

  for (n = 0; n < reso; n++) {
    if (state->T[n] > 230.1) {
      a[0] = -372.7;
      a[1] = 329.9;
      a[2] = 99.54;
      a[3] = 13.28;
      a[4] = 0.6449;
      b[0] = 1.898;
      b[1] = -1.68;
      b[2] = -0.5069;
      b[3] = -0.06758;
      b[4] = -0.003256;
    } else {
      a[0] = -61.72;
      a[1] = 54.64;
      a[2] = 16.48;
      a[3] = 2.198;
      a[4] = 0.1068;
      b[0] = 0.5479;
      b[1] = -0.485;
      b[2] = -0.1464;
      b[3] = -0.0195;
      b[4] = -0.00094;
    }
    A = a[0] + a[1] * x + a[2] * pow(x, 2.0) + a[3] * pow(x, 3.0) +
        a[4] * pow(x, 4.0);
    B = b[0] + b[1] * x + b[2] * pow(x, 2.0) + b[3] * pow(x, 3.0) +
        b[4] * pow(x, 4.0);
    radi[n] = A + B * state->T[n];
    if (radi[n] <= 0.0)
      radi[n] = 0.0;
  }
}

void calc_radi_s(const SimulationConfig *conf, const PlanetParams *para,
                 ClimateState *state, double *T, double radi_s[]) {
  int reso = conf->reso;
  int n;
  double x;
  double a[5], b[5];
  double A = 0.0, B = 0.0;
  double P_air = state->P_air;

  if (P_air > 7.45)
    P_air = 7.45;
  if (P_air < 3.4e-12)
    P_air = 3.4e-12;
  x = log10(P_air);

  for (n = 0; n < reso; n++) {
    if (T[n] > 230.1) {
      a[0] = -372.7;
      a[1] = 329.9;
      a[2] = 99.54;
      a[3] = 13.28;
      a[4] = 0.6449;
      b[0] = 1.898;
      b[1] = -1.68;
      b[2] = -0.5069;
      b[3] = -0.06758;
      b[4] = -0.003256;
    } else {
      a[0] = -61.72;
      a[1] = 54.64;
      a[2] = 16.48;
      a[3] = 2.198;
      a[4] = 0.1068;
      b[0] = 0.5479;
      b[1] = -0.485;
      b[2] = -0.1464;
      b[3] = -0.0195;
      b[4] = -0.00094;
    }
    A = a[0] + a[1] * x + a[2] * pow(x, 2.0) + a[3] * pow(x, 3.0) +
        a[4] * pow(x, 4.0);
    B = b[0] + b[1] * x + b[2] * pow(x, 2.0) + b[3] * pow(x, 3.0) +
        b[4] * pow(x, 4.0);
    radi_s[n] = A + B * T[n];
    if (radi_s[n] <= 0.0)
      radi_s[n] = 0.0;
  }
}

void calc_dif_s(const SimulationConfig *conf, const PlanetParams *para,
                ClimateState *state, double T_s[], double dif_s[]) {
  int reso = conf->reso;
  int n;
  double m = 10.0;

  for (n = 0; n < reso; n++) {
    dif_s[n] = m * (state->T[n] - T_s[n]);
  }
}

File: haskell-migration/ebm-mars/src/EBM/Atmosphere.hs
-- |
-- Module      : EBM.Atmosphere
-- Description : Atmospheric processes (diffusion and radiation)
-- Copyright   : (c) 2026
-- License     : MIT
--
-- Functions for calculating atmospheric diffusion and radiative transfer
-- on Mars, including both global and slope-specific calculations.
module EBM.Atmosphere
  ( calcDiffusion,
    calcRadiation,
    calcSlopeDiffusion,
    calcSlopeRadiation,
  )
where

import Data.Vector.Unboxed qualified as V
import EBM.Types

-- | Diffusion coefficient [bar^-1]
diffusionCoefficient :: Double
diffusionCoefficient = 5.3e-3

-- | Calculate atmospheric diffusion
calcDiffusion :: SimulationConfig -> PlanetParams -> ClimateState -> V.Vector Double
calcDiffusion config params state =
  let reso = resolution config
      d = diffusionCoefficient * unPressure (airPressure state)
      delPhi = pi / 180.0

      phi = V.generate reso (\n -> fromIntegral (n - 90) * pi / 180.0)
      temps = temperatures state
   in V.generate reso $ \n ->
        if n == 0 || n == reso - 1
          then 0.0 -- Boundary conditions
          else
            let phiN = phi V.! n
                tNext = temps V.! (n + 1)
                tPrev = temps V.! (n - 1)
                tCurr = temps V.! n
             in -d * tan phiN * (tNext - tPrev) * 0.5 / delPhi
                  + d * (tNext + tPrev - 2.0 * tCurr) / delPhi / delPhi

-- | Radiation coefficients for high temperature (T > 230.1 K)
highTempCoeffs :: ([Double], [Double])
highTempCoeffs =
  ( [-372.7, 329.9, 99.54, 13.28, 0.6449], -- a coefficients
    [1.898, -1.68, -0.5069, -0.06758, -0.003256] -- b coefficients
  )

-- | Radiation coefficients for low temperature (T <= 230.1 K)
lowTempCoeffs :: ([Double], [Double])
lowTempCoeffs =
  ( [-61.72, 54.64, 16.48, 2.198, 0.1068], -- a coefficients
    [0.5479, -0.485, -0.1464, -0.0195, -0.00094] -- b coefficients
  )

-- | Calculate radiation using polynomial coefficients
calcRadiationValue :: Double -> Double -> Double
calcRadiationValue pAir temp =
  let -- Clamp pressure to valid range
      pClamped =
        if pAir > 7.45
          then 7.45
          else
            if pAir < 3.4e-12
              then 3.4e-12
              else pAir

      x = log pClamped / log 10.0 -- log10

      -- Select coefficients based on temperature
      (aCoeffs, bCoeffs) =
        if temp > 230.1
          then highTempCoeffs
          else lowTempCoeffs

      -- Calculate A and B using polynomial
      calcPoly coeffs = sum $ zipWith (\c i -> c * x ** fromIntegral i) coeffs [0 ..]

      a = calcPoly aCoeffs
      b = calcPoly bCoeffs

      radi = a + b * temp
   in if radi <= 0.0 then 0.0 else radi

-- | Calculate atmospheric radiation
calcRadiation :: SimulationConfig -> PlanetParams -> ClimateState -> V.Vector Double
calcRadiation config params state =
  let reso = resolution config
      pAir = unPressure $ airPressure state
      temps = temperatures state
   in V.map (calcRadiationValue pAir) temps

-- | Calculate slope radiation
calcSlopeRadiation :: SimulationConfig -> PlanetParams -> ClimateState -> V.Vector Double -> V.Vector Double
calcSlopeRadiation config params state slopeTemps =
  let pAir = unPressure $ airPressure state
   in V.map (calcRadiationValue pAir) slopeTemps

-- | Calculate slope diffusion (heat exchange with atmosphere)
calcSlopeDiffusion :: SimulationConfig -> PlanetParams -> ClimateState -> V.Vector Double -> V.Vector Double
calcSlopeDiffusion config params state slopeTemps =
  let reso = resolution config
      m = 10.0 -- Exchange coefficient
      temps = temperatures state
   in V.zipWith (\tGlobal tSlope -> m * (tGlobal - tSlope)) temps slopeTemps
